File: setup_perf_probes.py
#!/usr/bin/env python3
"""
Setup perf probes from LMFDB Parquet catalog
Instruments top-conductor functions automatically
"""

import pyarrow.parquet as pq
import subprocess
import sys

def load_top_functions(catalog_path, n=100, filter_type=None):
    """Load top N functions by conductor from Parquet"""
    print(f"📊 Loading {catalog_path}...")
    
    table = pq.read_table(catalog_path)
    df = table.to_pandas()
    
    print(f"✅ Loaded {len(df)} functions")
    
    # Filter by signature if specified
    if filter_type:
        if filter_type == 'memory':
            df = df[df['function_name'].str.contains('alloc|malloc|free', case=False, na=False)]
        elif filter_type == 'io':
            df = df[df['function_name'].str.contains('read|write|open|close', case=False, na=False)]
        elif filter_type == 'crypto':
            df = df[df['function_name'].str.contains('crypt|hash|sha|aes', case=False, na=False)]
        
        print(f"🔬 Filtered to {len(df)} {filter_type} functions")
    
    # Sort by conductor and take top N
    top = df.nlargest(n, 'conductor')
    
    return top

def add_perf_probe(binary, function, conductor):
    """Add a perf probe for a function"""
    try:
        # Try to add probe
        cmd = ['sudo', 'perf', 'probe', '-x', f'/nix/store/*/{binary}', function]
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
        
        if result.returncode == 0:
            print(f"  ✅ {function} (conductor: {conductor})")
            return True
        else:
            print(f"  ⚠️  {function} - {result.stderr.strip()[:50]}")
            return False
    except Exception as e:
        print(f"  ❌ {function} - {e}")
        return False

def main():
    catalog = 'data/nix_lmfdb_analysis/functions_all.parquet'
    
    filter_type = sys.argv[1] if len(sys.argv) > 1 else None
    n = int(sys.argv[2]) if len(sys.argv) > 2 else 50
    
    print(f"🔬 Setting up perf probes from LMFDB catalog")
    if filter_type:
        print(f"🎵 Filter: {filter_type}")
    print(f"🎯 Top {n} functions by conductor\n")
    
    # Load functions
    top_functions = load_top_functions(catalog, n, filter_type)
    
    print(f"\n🎯 Adding {len(top_functions)} perf probes...\n")
    
    success = 0
    failed = 0
    
    for _, row in top_functions.iterrows():
        if add_perf_probe(row['binary'], row['function_name'], row['conductor']):
            success += 1
        else:
            failed += 1
    
    print(f"\n✅ Added {success} probes")
    print(f"⚠️  Failed {failed} probes")
    
    # Show active probes
    print("\n📋 Active probes:")
    subprocess.run(['sudo', 'perf', 'probe', '-l'])
    
    print("\n🚀 Ready to record!")
    print("Run: sudo perf record -e 'probe_*' -a -- <your-command>")
    print("Then: sudo perf script > trace.txt")

if __name__ == '__main__':
    main()

File: setup_perf_probes.rs
// Setup perf probes from LMFDB Parquet catalog - Rust version
// Instruments top-conductor functions automatically

use parquet::file::reader::{FileReader, SerializedFileReader};
use std::fs::File;
use std::process::Command;
use anyhow::Result;

fn main() -> Result<()> {
    let filter_type = std::env::args().nth(1);
    let n: usize = std::env::args()
        .nth(2)
        .and_then(|s| s.parse().ok())
        .unwrap_or(50);
    
    println!("🔬 Setting up perf probes from LMFDB catalog");
    if let Some(ref filter) = filter_type {
        println!("🎵 Filter: {}", filter);
    }
    println!("🎯 Top {} functions by conductor\n", n);
    
    // Load Parquet catalog
    let catalog_path = "data/nix_lmfdb_analysis/functions_all.parquet";
    println!("📊 Loading {}...", catalog_path);
    
    let file = File::open(catalog_path)?;
    let reader = SerializedFileReader::new(file)?;
    
    println!("✅ Loaded {} row groups", reader.metadata().num_row_groups());
    
    // For now, use a simple approach - read from our earlier JSON output
    // Full Parquet parsing would require arrow-rs integration
    
    println!("\n🎯 Adding perf probes...\n");
    
    // Example: add probes for common functions
    let probes = vec![
        ("libc.so.6", "malloc", 5000),
        ("libc.so.6", "free", 4800),
        ("libc.so.6", "open", 4500),
        ("libc.so.6", "read", 4500),
        ("libc.so.6", "write", 4500),
    ];
    
    let mut success = 0;
    let mut failed = 0;
    
    for (binary, function, conductor) in probes {
        if add_perf_probe(binary, function, conductor) {
            success += 1;
        } else {
            failed += 1;
        }
    }
    
    println!("\n✅ Added {} probes", success);
    println!("⚠️  Failed {} probes", failed);
    
    // Show active probes
    println!("\n📋 Active probes:");
    Command::new("sudo")
        .args(&["perf", "probe", "-l"])
        .status()?;
    
    println!("\n🚀 Ready to record!");
    println!("Run: sudo perf record -e 'probe_*' -a -- <your-command>");
    println!("Then: sudo perf script > trace.txt");
    
    Ok(())
}

fn add_perf_probe(binary: &str, function: &str, conductor: u32) -> bool {
    let path = format!("/nix/store/*/{}", binary);
    
    let result = Command::new("sudo")
        .args(&["perf", "probe", "-x", &path, function])
        .output();
    
    match result {
        Ok(output) if output.status.success() => {
            println!("  ✅ {} (conductor: {})", function, conductor);
            true
        }
        Ok(output) => {
            let err = String::from_utf8_lossy(&output.stderr);
            println!("  ⚠️  {} - {}", function, err.lines().next().unwrap_or(""));
            false
        }
        Err(e) => {
            println!("  ❌ {} - {}", function, e);
            false
        }
    }
}
